[JSC] Use TerminateProcess for Windows

Constellation opened the pull request:
#### 4c277ef7513887fe3c335b087572cb30bd1aec9f
<pre>
[JSC] Use TerminateProcess for Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=254566">https://bugs.webkit.org/show_bug.cgi?id=254566</a>
rdar://107296359

Reviewed by Mark Lam.

Forcefully terminate the process via TerminateProcess on Windows at the end of main function in JSC shell.
We are running concurrent collectors and normal exit procedure can cause dead lock on Windows since Windows
does not really have &quot;main thread&quot; concept: on the other platforms, when main thread dies, everything gets
shut down. But on Windows, it is not.

* Source/JavaScriptCore/jsc.cpp:
(JSC::jscExit):

Canonical link: <a href="https://commits.webkit.org/262193@main">https://commits.webkit.org/262193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc33ad123d08ffb3da154a5b8a34fd3a3a3f357e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/871 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/895 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/930 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/1231 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/769 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/945 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/980 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/981 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/879 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/820 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1170 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/870 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/812 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/814 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/765 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/787 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/838 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/1828 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/854 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/828 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/895 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/774 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/823 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/188 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/841 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/898 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/92 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/182 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -194,7 +194,11 @@ NO_RETURN_WITH_VALUE static void jscExit(int status)
     }
 #endif // ENABLE(DFG_JIT)
+#if OS(WINDOWS)
+    TerminateProcess(GetCurrentProcess(), status);
```
You might need to `#include <processthreadsapi.h>`?

Constellation replied:
I guess `windows.h` is enough, but let's see the build on EWS. And if it fails, then let's add it.

Constellation replied:
WinCairo build bot get built. So yeah, `windows.h` was enough. Landing.

Constellation committed:
[JSC] Use TerminateProcess for Windows
https://bugs.webkit.org/show_bug.cgi?id=254566
rdar://107296359

Reviewed by Mark Lam.

Forcefully terminate the process via TerminateProcess on Windows at the end of main function in JSC shell.
We are running concurrent collectors and normal exit procedure can cause dead lock on Windows since Windows
does not really have "main thread" concept: on the other platforms, when main thread dies, everything gets
shut down. But on Windows, it is not.

* Source/JavaScriptCore/jsc.cpp:
(JSC::jscExit):

Canonical link: https://commits.webkit.org/262193@main
